dev/io.openliberty.microprofile.telemetry.1.0.internal/src/io/openliberty/microprofile/telemetry/internal/cdi/WithSpanInterceptor.java

```diff
@@ -75,52 +86,63 @@ private static SpanKind spanKindFromMethod(Method method) {
-        MethodRequest methodRequest = new MethodRequest(invocationContext.getMethod(), invocationContext.getParameters());
-
-        Context parentContext = Context.current();
-        Context spanContext = null;
-        Scope scope = null;
-        boolean shouldStart = instrumenter.shouldStart(parentContext, methodRequest);
-        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
-            Tr.debug(this, tc, "Method " + invocationContext.getMethod().toString() + " Should start: " + shouldStart);
-        }
+        try {
+            MethodRequest methodRequest = new MethodRequest(invocationContext.getMethod(), invocationContext.getParameters());
 
-        if (shouldStart) {
-            spanContext = instrumenter.start(parentContext, methodRequest);
-            scope = spanContext.makeCurrent();
+            Context parentContext = Context.current();
+            Context spanContext = null;
+            Scope scope = null;
+            boolean shouldStart = instrumenter.shouldStart(parentContext, methodRequest);
             if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
-                Tr.debug(this, tc, "spanContext " + spanContext.toString() + " has started and is now the current context");
+                Tr.debug(this, tc, "Method " + invocationContext.getMethod().toString() + " Should start: " + shouldStart);
             }
-        }
 
-        try {
-            Object result = invocationContext.proceed();
             if (shouldStart) {
-                instrumenter.end(spanContext, methodRequest, null, null);
+                spanContext = instrumenter.start(parentContext, methodRequest);
+                scope = spanContext.makeCurrent();
                 if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
-                    Tr.debug(this, tc, "spanContext " + spanContext.toString() + " has ended");
+                    Tr.debug(this, tc, "spanContext " + spanContext.toString() + " has started and is now the current context");
                 }
             }
-            return result;
-        } finally {
-            if (scope != null) {
-                scope.close();
+
+            try {
+                Object result = invocationContext.proceed();
+                if (shouldStart) {
+                    instrumenter.end(spanContext, methodRequest, null, null);
+                    if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
+                        Tr.debug(this, tc, "spanContext " + spanContext.toString() + " has ended");
+                    }
+                }
+                return result;
+            } finally {
+                if (scope != null) {
+                    scope.close();
+                }
             }
-        }
+        } catch (Exception e) {
+            Tr.error(tc, Tr.formatMessage(tc, "CEMOT5101.telemetry.error", e));
+            return null;
+        } 
```

Comment:
This gets a bit tricky. We're an interceptor, so we should return the result of `invocationContext.proceed()` - but at this point we don't know if we've already called it or not.

Reply:
> I think the right thing to do is probably to catch any exception from proceed() separately (without FFDC), pass it to instrumentation.end() and rethrow it at the end of the method.

Why not just call instrumentation.end() from the finally block? There is a try that only contains proceed() and some logging?

Reply:
Ah, sorry, there's a separate bug in the original code.

If the method returns a value, that value should be passed to `instrumentation.end()`. If the method throws an exception, that exception should be passed to `instrumentation.end()` instead. [Javadoc here](https://www.javadoc.io/doc/io.opentelemetry.instrumentation/opentelemetry-instrumentation-api/1.19.0/io/opentelemetry/instrumentation/api/instrumenter/Instrumenter.html)

This would be a release bug I think and needs a test, so maybe we should do it in a separate PR.